packages/playground/website/src/components/playground-viewport/index.tsx

```diff
@@ -21,52 +28,110 @@ interface PlaygroundViewportProps {
 	className?: string;
 }
 
-export const PlaygroundViewport = React.forwardRef<
-	HTMLIFrameElement,
-	PlaygroundViewportProps
->(
-	(
-		{ displayMode = 'browser', toolbarButtons, hideToolbar, className },
-		ref
-	) => {
-		const { playground, currentUrl: url } = usePlaygroundContext();
+export const PlaygroundViewport = ({
+	displayMode = 'browser-full-screen',
+	toolbarButtons,
+	hideToolbar,
+	className,
+}: PlaygroundViewportProps) => {
+	const clientInfo = useAppSelector(getActiveClient);
+	const playground = clientInfo?.client;
+	const url = clientInfo?.url;
 
-		if (displayMode === 'seamless') {
-			// No need to boot the playground if seamless.
-			return <JustViewport iframeRef={ref} />;
-		}
-		return (
-			<BrowserChrome
-				showAddressBar={!!playground}
-				url={url}
-				toolbarButtons={toolbarButtons}
-				onUrlChange={(url) => playground?.goTo(url)}
-				hideToolbar={hideToolbar}
-				className={className}
-			>
-				<JustViewport iframeRef={ref} />
-			</BrowserChrome>
-		);
+	if (displayMode === 'seamless') {
+		// No need to boot the playground if seamless.
+		return <JustViewport />;
 	}
-);
+	return (
+		<BrowserChrome
+			showAddressBar={!!playground}
+			url={url}
+			toolbarButtons={toolbarButtons}
+			onUrlChange={(url) => playground?.goTo(url)}
+			hideToolbar={hideToolbar}
+			className={className}
+		>
+			<JustViewport />
+		</BrowserChrome>
+	);
+};
 
-interface JustViewportProps {
-	iframeRef: Ref<HTMLIFrameElement>;
-}
+export const JustViewport = function LoadedViewportComponent() {
+	const iframeRef = useRef<HTMLIFrameElement>(null);
+	const iframe = (iframeRef as any)?.current;
+	useEffect(() => {
+		if (iframe) {
+			setupPostMessageRelay(iframe, document.location.origin);
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [!!iframe]);
```

Comment:
And if it is correct that we want to do new setup any time the iframe instance changes, we may also need to remove the listeners that refer to the previous iframe instance.